### PR TITLE
Remove incorrect vmIDsToWindowHandles cache

### PR DESCRIPTION
**Pull request by rpaquay**
The vmIDsToWindowHandles cache is base on an incorrect assumption
that there is a mapping between a jvmid to a _single_ Window Handle.
This is incorrect as a Java application can have many top level
windows (each top level frame corresponds to a unique Window Handle).

The cache was introduced in 2007:

https://github.com/nvaccess/nvda/commit/d4a74c7cf155cedcd1cf48b47604a9c5d667e83b

This code was incorrect at the time, but was good enough to make things
work with Java applications that have only one top level frame.

A commit in 2010 implemented the correct way of retrieving the window
handle from a given AccessibleContext:

https://github.com/nvaccess/nvda/commit/eddfb4bd3e193ae4145dd6a697babc4bc12582e5

This effectively removed the need of using cache, but the cache was
not removed for some reason.

The symptom of the issue is described in bug #5732.

### CODE REVIEW

**Comment by jcsteh** [comment]
@MichaelDCurran, you know more about the history of this code than I, so 
it'd be good if you can comment. This patch makes total sense to me 
given our incorrect assumption. Can you think of any reason this would 
cause problems?

**Comment by michaelDCurran** [comment]
The logic sounds fine to me. I guess that getTopLevelObject or getHWNDFromAccessibleContext may not have been implemented in java Access Bridge 1.x or something. As long as we are happy assuming that everyone is using 2 now, then  it sounds good to me.

**Comment by rpaquay** [comment]
Hmm. Hold on onto this patch. With this patch, I am seeing cases where "getTopLevelObject" fails intermittently, mainly when nvda processes "focus" events. The failure does not seems to happen with "simple" apps (e.g. SwingSet2), but it does happen with some of the various experimental changes I have been working on to try to improve the accessibility behavior of IntelliJ's Code Completion feature. I am investigating the root cause, but I think overall, it might be safer to keep the cache, but use it only as fallback when "getTopLevelObject" fails. I will update the PR later.

**Comment by rpaquay** [comment]
@MichaelDCurran @jcsteh 

I tweaked the PR to keep the cache, but use it only as a fallback when getTopLevelObject fails. This improve the existing behavior, although it is still not 100% reliable.

I also changed the JAB(window)._isEqual method to only compare jabContext for equality, ignoring window handles (since we can't make them reliable). This is similar to how UIA(Window)._isEqual is implemented.

I understand all of this change looks a little scary and abstract, but this improves how nvda handles a succession of quick "focus gained" events that switch between two top level windows of a given JVM ID. This is helpful to help me fix a bug in Android Studio/IntelliJ: https://code.google.com/p/android/issues/detail?id=198011

In any ways, even though the change helps me with a specific bug in AS/IntelliJ, I think the change is for the better in general, as we all seem to agree that a cache from JVM ID to HWND is based on an incorrect assumption. Let me know what you guys think. Thanks!

**Comment by jcsteh** [comment]
I agree; I think this change is for the better. I'd prefer we had a better understanding of why getTopLevelObject, etc. sometimes fails, but we may never know and it's very probably due to some obscure bug in the mostly-abandoned JAB.

I have a few questions, but I'll ask those inline.

Btw, ignore the CI failure on this PR. It isn't intended to build pull requests. I've fixed this now.

File: source/JABHandler.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#javaAccessBridgeHandler.py
+﻿#javaAccessBridgeHandler.py
```
Comment (jcsteh) [inline source/JABHandler.py]: nit: A Unicode BOM has been added here. Frustratingly, this breaks xgettext. Please remove it. I'll update the Contributing document to mention this.

File: source/JABHandler.py [hunk 2]
```diff
@@ -519,11 +539,12 @@ def getAccessibleKeyBindings(self):
...
 @AccessBridge_FocusGainedFP
 def internal_event_focusGained(vmID, event,source):
-	internalQueueFunction(event_gainFocus,vmID,source)
+	hwnd=getWindowHandleFromAccContext(vmID,source)
+	internalQueueFunction(event_gainFocus,vmID,source,hwnd)
 	bridgeDll.releaseJavaObject(vmID,event)
 
-def event_gainFocus(vmID,accContext):
-	jabContext=JABContext(vmID=vmID,accContext=accContext)
+def event_gainFocus(vmID,accContext,hwnd):
```
Comment (jcsteh) [inline source/JABHandler.py]: I'm probably missing something obvious here, but I don't understand why you've added hwnd arguments to the event_* functions. From what I can see, you always call getWindowHandleFromAccContext to get the hwnd before calling the event, but JABContext will do this anyway if there's no hwnd provided. Can you please explain?

File: source/NVDAObjects/JAB/__init__.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-import ctypes
+﻿import ctypes
```
Comment (jcsteh) [inline source/NVDAObjects/JAB/__init__.py]: nit: Please remove Unicode BOM.

File: source/NVDAObjects/JAB/__init__.py [hunk 4]
```diff
@@ -229,7 +229,12 @@ def _get_TextInfo(self):
 		return super(JAB,self).TextInfo
 
 	def _isEqual(self,other):
-		return super(JAB,self)._isEqual(other) and self.jabContext==other.jabContext
+		if not isinstance(other,JAB):
+			return False
```
Comment (jcsteh) [inline source/NVDAObjects/JAB/__init__.py]: nit: This check is redundant, as `NVDAObject.__eq__` (which calls _isEqual) already checks that the types are the same.

File: source/JABHandler.py (hunk 2 above)
Comment (rpaquay) [inline source/JABHandler.py]: There are 2 reasons:

* Main reason: In the enterJavaWindow_helper function, we get a valid hwnd and we use that to enqueue the "event_gainFocus" event. This means that in that case, we have a valid hwnd 100% of the time when processing "event_fainFocus". In other word, this is a code path where we don't need to call "getWindowHandleFromAccContext" to obtain a hwnd.

* Minor reason: By acquiring the hwnd at the time the event is enqueued, as opposed to the time the event is processed, we improve the chances of obtaining a valid hwnd from the AccessibleContext. The cases where I have had issues with unreliable hwnd have been related to focus switching quickly between 2 windows, so there is a timing aspect in this.

File: source/NVDAObjects/JAB/__init__.py (hunk 4 above)
Comment (rpaquay) [inline source/NVDAObjects/JAB/__init__.py]: Ok, I will remove it -- Note I copied the implementation of UIA(Window), which has the extra check too. What do you think of the "try/except" around the jabcontext comparison. Is it necessary?

File: source/JABHandler.py (hunk 1 above)
Comment (rpaquay) [inline source/JABHandler.py]: Will do. I think my text editor (Visual Studio) does that automatically :(

File: source/NVDAObjects/JAB/__init__.py (hunk 3 above)
Comment (rpaquay) [inline source/NVDAObjects/JAB/__init__.py]: Will do. I think my text editor (Visual Studio) does that automatically :(

**Comment by rpaquay** [comment]
@MichaelDCurran @jcsteh

The PR should be ready to go in as I think I addressed all comments. Let me know if there is anything else I need to do. Thanks!

File: source/NVDAObjects/JAB/__init__.py (hunk 4 above)
Comment (jcsteh) [inline source/NVDAObjects/JAB/__init__.py]: > Ok, I will remove it -- Note I copied the implementation of UIA(Window), which has the extra check too.

Oh, sorry, so it does. That one is redundant too. :(

> What do you think of the "try/except" around the jabcontext comparison. Is it necessary?

`JABContext.__eq__` really shouldn't ever throw an exception--it calls isSameObject, which doesn't report errors--but I guess having the try/except there probably doesn't hurt as far as being future proof.

**Pull request closed without merging**